Ship of Harkinian 64 bit compilation

KiritoDv opened the pull request.

MelonSpeedruns committed:
Fixed Gyroscopy Settings & Drift

MelonSpeedruns committed:
Merge branch 'HarbourMasters:develop' into develop

MelonSpeedruns committed:
Merge branch 'HarbourMasters:develop' into develop

MelonSpeedruns committed:
Merge branch 'HarbourMasters:develop' into develop

Emill committed:
64-bit early draft - no audio yet

Emill committed:
Fix VS stuff

sholdee committed:
Update platform for z64 branch (#338)

KiritoDv committed:
Fixed soh filters

Kenix3 committed:
Audio code converted to uintptr_t

Kenix3 committed:
Merge branch 'zelda64' of https://github.com/HarbourMasters/Shipwright into zelda64

MelonSpeedruns committed:
Merge branch 'HarbourMasters:develop' into develop

sholdee committed:
Update soh path for z64 branch (#344)

This check will fail on Windows before it gets merged btw

KiritoDv committed:
Fixed camera and disabled sound

KiritoDv committed:
Merge branch 'zelda64' of https://github.com/HarbourMasters/Shipwright into zelda64

KiritoDv committed:
Uncommented item_24 print on z_message_PAL

KiritoDv committed:
Fixed z_player_lib corruption on 64 bit

KiritoDv committed:
Fixed oob crashes on 64 bit

KiritoDv committed:
Merge develop into zelda64

louist103 commented on libultraship/libultraship/Lib/Fast3D/U64/PR/ultra64/gbi.h:
```diff
@@ -1637,13 +1637,19 @@ typedef struct {
 } Gwords;
 
+#ifdef __cplusplus
+static_assert(sizeof(Gwords) == 2 * sizeof(void*), "fail");
```
Should put something descriptive in here.

Kenix3 replied:
```suggestion
static_assert(sizeof(Gwords) == 2 * sizeof(void*), "Display list size is bad");
```

louist103 commented on soh/soh/OTRGlobals.h:
```diff
@@ -37,18 +37,22 @@ uint32_t ResourceMgr_GetGameVersion();
 CollisionHeader* ResourceMgr_LoadColByName(const char* path);
 uint64_t GetPerfCounter();
 struct SkeletonHeader* ResourceMgr_LoadSkeletonByName(const char* path);
+s32* ResourceMgr_LoadCSByName(const char* path);
```
Should return a pointer to the union

Kenix3 replied:
This is more of a documentation change than a change to make 64 bit work, and thus out of the scope of this PR.

louist103 commented on soh/src/code/z_bgcheck.c:
```diff
@@ -1516,7 +1516,7 @@ void BgCheck_Allocate(CollisionContext* colCtx, GlobalContext* globalCtx, Collis
             // "/* BGCheck LonLon Size %dbyte */\n"
             osSyncPrintf("/* BGCheck LonLonサイズ %dbyte */\n", 0x3520);
-            colCtx->memSize = 0x3520;
+            colCtx->memSize = 0x3520 * (sizeof(void*) / 4); // 64-bit build needs more memory
```
This should probably be a macro with defines for 32 or 64 bit.

NEstelami committed:
Audio decompiled and WIP custom sample support

NEstelami committed:
Audio decompiled and WIP custom sample support

NEstelami committed:
Audio decompiled and WIP custom sample support

NEstelami committed:
Replaced StormLib .lib files with source code.

Some modifications to handle backslashes and forward slashes, along with some optimizations to speed up OTR generation.

NEstelami committed:
Replaced StormLib .lib files with source code.

Some modifications to handle backslashes and forward slashes, along with some optimizations to speed up OTR generation.

NEstelami committed:
Replaced StormLib .lib files with source code.

Some modifications to handle backslashes and forward slashes, along with some optimizations to speed up OTR generation.

KiritoDv committed:
Merge branch 'zapd_audio_support' into zelda64

KiritoDv committed:
Merge branch 'zapd_audio_support' into zelda64

dcvz committed:
Fix function prototypes and types

KiritoDv committed:
Added missing function on OTRGlobals

KiritoDv committed:
Fixed audio font loading

dcvz committed:
Add StormLib CMakeLists.txt for building on Linux/Mac

KiritoDv committed:
Increased initial audio pool size and fixed types

KiritoDv committed:
Disabled RTC on soh and lus

Random06457 committed:
fix compilation

Random06457 committed:
change OSMesg definition and fix memory corruptions

Random06457 committed:
fix Makefile typo

qurious-pixel commented on soh/Makefile:
```diff
@@ -54,7 +56,7 @@ INC_DIRS := $(addprefix -I, \
 )
 
 LDDIRS := $(addprefix -L, \
-	../external \
```
On my build wasn't able to find libGLEW.a in `/usr/local/lib`
Copied it into StormLib/build and the linking succeeded.

Is there a better way to find GLEW?

Emill replied:
Actually on Linux I'm not sure we even need glew. Usually it is possible to directly link to libOpenGL.

qurious-pixel replied:
Does that include RayLib?

Emill replied:
I have no idea how raylib is set up, but I assume it applies there as well.

Kenix3 reviewed (commented):
This works, right? Any specific reason this PR is still flagged as draft? I believe it's ready for review.

Emill commented on soh/include/z64.h:
```diff
@@ -30,7 +30,8 @@
 
-#define AUDIO_HEAP_SIZE 0x38000
+//#define AUDIO_HEAP_SIZE 0x38000
+#define AUDIO_HEAP_SIZE 0x3800000 // The original audio heap size was too small. The heap would exceed capacity and corrupt everything in its path.
```
Does it really have to be 256 times larger?

Emill commented on soh/include/z64camera.h:
```diff
@@ -1157,7 +1157,7 @@ typedef struct {
 
 typedef struct {
-    /* 0x000 */ char paramData[0x50];
+    /* 0x000 */ char paramData[0xFF];
```
`0x14 * sizeof(void*)`?

Emill commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -530,20 +539,247 @@ extern "C" CollisionHeader* ResourceMgr_LoadColByName(const char* path)
+
+    if (strem2[0] == 'R' && strem2[1] == 'I' && strem2[2] == 'F' && strem2[3] == 'F')
+    {
+        SoundFontSample* sampleC = (SoundFontSample*)malloc(sizeof(SoundFontSample));
```
We can use `new SoundFontSample` instead of malloc since we use C++. That will either throw an error or terminate rather than returning null (which is currently not handled).

Emill commented on soh/src/code/audio_init_params.c:
```diff
@@ -10,7 +10,7 @@ const s16 D_8014A6C0[] = {
 };
 
-const AudioContextInitSizes D_8014A6C4 = { 0x37F00, 0xE0E0, 0xBCE0 };
+const AudioContextInitSizes D_8014A6C4 = { 0x57F00, 0x2E0E0, 0xFCE0 };
```
Any reason to go from 0xe0e0 to 0x2e0e0, which is quite a big increase?

Emill commented on soh/src/code/audio_load.c:
```diff
@@ -665,29 +702,55 @@ SoundFontData* AudioLoad_SyncLoadFont(u32 fontId) {
-        medium = table->entries[id].medium;
-        cachePolicy = table->entries[id].cachePolicy;
-        romAddr = table->entries[realId].romAddr;
+    } else
+    {
```
Try to use the same coding style as the file being edited already uses.

Emill commented on soh/src/code/z_kanfont.c:
```diff
@@ -183,7 +183,7 @@ void Font_LoadChar(Font* font, u8 character, u16 codePointIndex) {
  */
 void Font_LoadMessageBoxIcon(Font* font, u16 icon) {
-    memcpy(font->iconBuf, msgStaticTbl[4 + icon], FONT_CHAR_TEX_SIZE);
+    memcpy(font->iconBuf, ResourceMgr_LoadTexByName(msgStaticTbl[4 + icon]), FONT_CHAR_TEX_SIZE);
```
strcpy or something could work as well

louist103 replied:
`strcpy` will stop at the first zero byte.

Emill replied:
Yes? It's a string being copied so that's what we want.

louist103 replied:
This is copying the texture into the icon buffer not the string for the name.

Emill commented on soh/src/code/z_lifemeter.c:
```diff
@@ -162,9 +162,9 @@ void HealthMeter_Init(GlobalContext* globalCtx) {
-    sHeartsDDPrim[2][0] = HeartInner[0];
-    sHeartsDDPrim[2][1] = HeartInner[1];
-    sHeartsDDPrim[2][2] = HeartInner[2];
+    // sHeartsDDPrim[2][0] = HeartInner[0];
+    // sHeartsDDPrim[2][1] = HeartInner[1];
+    // sHeartsDDPrim[2][2] = HeartInner[2];
```
Why are these removed?

Random06457 replied:
I removed these because they are OOB writes (if you look at `sHeartsDDPrim` it's defined as `sHeartsDDPrim[2][3]`). I don't know what the intentions of the person that added these were but I assumed it was a mistake here.

Emill commented on soh/src/overlays/actors/ovl_Boss_Goma/z_boss_goma.c:
```diff
@@ -326,6 +326,7 @@ void BossGoma_ClearPixels32x32Rgba16(s16* rgba16image, u8* clearPixelTable, s16
  * Clear pixels from Gohma's textures
  */
 void BossGoma_ClearPixels(u8* clearPixelTable, s16 i) {
+    return;
```
Can this be fixed somehow?

dcvz replied:
It's actually the `BossGoma_ClearPixels32x32Rgba16` method that's crashing. I think we should be able to rewrite that so it also works on 64bit machines

NEstelami committed:
Audio support nearly complete.

NEstelami committed:
Audio support nearly complete.

NEstelami committed:
Audio support nearly complete.

NEstelami committed:
Commented out NTSC test

NEstelami committed:
Commented out NTSC test

NEstelami committed:
Commented out NTSC test

NEstelami committed:
Commented out additional test code

NEstelami committed:
Commented out additional test code

NEstelami committed:
Commented out additional test code

NEstelami committed:
Fixed resample and corrupt audio bug. Minor documentation updates.

NEstelami committed:
Fixed resample and corrupt audio bug. Minor documentation updates.

NEstelami committed:
Fixed resample and corrupt audio bug. Minor documentation updates.

dcvz commented:
@KiritoDv is there still some open tasks left? This is still marked as a draft :)

KiritoDv committed:
Code cleanup and fixes on 64bits

KiritoDv committed:
Fixed conflicts with develop

KiritoDv committed:
Updated ResourceMgr to develop changes

KiritoDv committed:
Merge branch 'develop' into zelda64

KiritoDv committed:
Updated zapd_audio_support on zelda64

KiritoDv committed:
Moved .o files into x64/DebugObj

KiritoDv committed:
Moved .o files to DebugObj - Debug

dcvz committed:
Fix detection of arm64, fix warnings

KiritoDv committed:
Changed audio freq from 32KHZ to 44KHZ on SDLAudioPlayer

NEstelami committed:
Fixed crackling issues

NEstelami committed:
Fixed crackling issues

NEstelami committed:
Fixed crackling issues

NEstelami committed:
Updated sample rate to 44100hz

NEstelami committed:
Updated sample rate to 44100hz

NEstelami committed:
Updated sample rate to 44100hz

KiritoDv committed:
Merge branch 'zapd_audio_support' into zelda64

NEstelami committed:
Fixed last remaining audio bugs

NEstelami committed:
Fixed last remaining audio bugs

NEstelami committed:
Fixed last remaining audio bugs

KiritoDv committed:
Fixed ocarina loop points

NEstelami committed:
Added names for the soundfonts.

NEstelami committed:
Added names for the soundfonts.

NEstelami committed:
Added XML attributes and fixed odd crash

NEstelami committed:
Added XML attributes and fixed odd crash

dcvz commented on libultraship/libultraship/SDLAudioPlayer.cpp:
```diff
@@ -9,7 +9,7 @@ namespace Ship {
             SDL_AudioSpec want, have;
             SDL_zero(want);
-            want.freq = 32000;
+            want.freq = 44000;
```
This should be 44100

dcvz commented:
@NEstelami @KiritoDv we probably want to update PulseAudioPlayer’s rate variable as well

NEstelami committed:
GC PAL Audio XML

NEstelami committed:
GC PAL Audio XML

KiritoDv committed:
Added GetSampleRate function

NEstelami committed:
Fixed StormLib Forward Slash BackSlash issues

NEstelami committed:
Fixed StormLib Forward Slash BackSlash issues

KiritoDv committed:
Fixed Ganondorf's lair and Spirit Temple crashes on 64 bits

KiritoDv committed:
Merge branch 'develop' into zelda64

KiritoDv committed:
Updated zapd_audio_support into zelda64

KiritoDv committed:
Disabled ASAN on release

KiritoDv committed:
Fixed x64 release compilation

NEstelami committed:
Updated sequence file format and updated resource version to Rachael

KiritoDv committed:
Added mp3 support for samples

KiritoDv committed:
Code cleanup

KiritoDv committed:
Fixed indentation caused because of tabs

NEstelami committed:
Updated resource types

NEstelami committed:
Fixed Linux issues (I think)

NEstelami committed:
Let's try again

NEstelami committed:
Ready for round 2?

NEstelami committed:
Additional fixes

KiritoDv committed:
Merge branch 'zapd_audio_support' into zelda64

Commit:
Merge branch 'develop' into z64-devmerge

NEstelami committed:
Merge pull request #494 from briaguya-ai/z64-devmerge

Z64 devmerge

dcvz commented:
@NEstelami @KiritoDv @Emill @Kenix3 If there are no other major blockers, how do you all feel about merging in this branch and fixing any other minor issues that appear separately? This PR has a lot of useful and helpful additions contained only within this branch that would enable a couple of other efforts + it would allow more people to be able to test and contribute to the stabilization of `develop` and 64bit.

Or if not, is there anything we can do to help with the effort? is there a list of bugs somewhere that we (other devs) can help track down and fix? :)

Kenix3 commented:
> @NEstelami @KiritoDv @Emill @Kenix3 If there are no other major blockers, how do you all feel about merging in this branch and fixing any other minor issues that appear separately? This PR has a lot of useful and helpful additions contained only within this branch that would enable a couple of other efforts + it would allow more people to be able to test and contribute to the stabilization of `develop` and 64bit.
> 
> Or if not, is there anything we can do to help with the effort? is there a list of bugs somewhere that we (other devs) can help track down and fix? :)

Just a couple days ago it finally got some code submissions that were blocking it to be merged. Give it a few days for review and it will get in.

Kenix3 commented on soh/src/code/z_bgcheck.c:
```diff
@@ -1584,6 +1584,11 @@ void BgCheck_Allocate(CollisionContext* colCtx, GlobalContext* globalCtx, Collis
                                     &colCtx->subdivLength.y, &colCtx->subdivLengthInv.y);
     BgCheck_SetSubdivisionDimension(colCtx->minBounds.z, colCtx->subdivAmount.z, &colCtx->maxBounds.z,
                                     &colCtx->subdivLength.z, &colCtx->subdivLengthInv.z);
+
```
This seems super hacky. Why an ifdef? Can't we just calculate the same for 32 and 64 bit?

KiritoDv replied:
Nope because it crashes on some maps, i could replace to a x2 if you want

Kenix3 reviewed (commented):
First, has the 32 bit build been tested on this branch at all?

Second, why is SoH building StormLib in the makefile, but not libultraship? StormLib is only included to SoH via libultraship, so the logical place to build it would be libultraship.

Kenix3 committed:
Update libultraship/libultraship/Lib/Fast3D/U64/PR/ultra64/gbi.h

dcvz commented:
> > @NEstelami @KiritoDv @Emill @Kenix3 If there are no other major blockers, how do you all feel about merging in this branch and fixing any other minor issues that appear separately? This PR has a lot of useful and helpful additions contained only within this branch that would enable a couple of other efforts + it would allow more people to be able to test and contribute to the stabilization of `develop` and 64bit.
> > Or if not, is there anything we can do to help with the effort? is there a list of bugs somewhere that we (other devs) can help track down and fix? :)
> 
> Just a couple days ago it finally got some code submissions that were blocking it to be merged. Give it a few days for review and it will get in.

No rush. Was just asking to avoid having a big PR that gets harder to rebase with an ever changing `develop` and to avoid isolating some functionality that could benefit mainline (like additional audio improvements). As SoH already advertises that nightly builds are unstable I thought it might be okay.

I'm also just genuinely trying to help. I know moving such big efforts forward can be a big and tiring task and right now it seems that that responsibility is solely falling on one or two people.

louist103 commented on libultraship/libultraship/DisplayList.cpp:
```diff
@@ -12,20 +12,51 @@ namespace Ship
 			reader->ReadByte();
 
-		while (true)
+		if (sizeof(uintptr_t) < 8)
```
Seems like something we should `ifdef`

louist103 commented on libultraship/libultraship/Lib/Fast3D/U64/PR/ultra64/gbi.h:
```diff
@@ -1637,13 +1637,19 @@ typedef struct {
 
+#ifdef __cplusplus
+static_assert(sizeof(Gwords) == 2 * sizeof(void*), "Display list size is bad");
+#else
+//_Static_assert(sizeof(Gwords) == 2 * sizeof(void*), "fail");
```
I's say this can go. or create a C error check

louist103 commented on libultraship/libultraship/PulseAudioPlayer.cpp:
```diff
@@ -74,7 +74,7 @@ namespace Ship
         pa_sample_spec ss;
         ss.format = PA_SAMPLE_S16LE;
-        ss.rate = 32000;
+        ss.rate = this->GetSampleRate();
```
as much as I hate it we don't really use `this` unless we have to

louist103 commented on libultraship/libultraship/PulseAudioPlayer.cpp:
```diff
@@ -149,8 +149,7 @@ namespace Ship
 
     int PulseAudioPlayer::GetDesiredBuffered()
```
can't this be a `constexpr` like the other one?

louist103 commented on libultraship/libultraship/SDLAudioPlayer.cpp:
```diff
@@ -2,40 +2,40 @@
             if (Device == 0) {
-                SPDLOG_ERROR("SDL_OpenAudio error: {}", SDL_GetError());
-                return false;
+                  SPDLOG_ERROR("SDL_OpenAudio error: {}", SDL_GetError());
+                  return false;
```
looks like your spaces got weird here.

louist103 commented on soh/include/z64.h:
```diff
@@ -30,8 +30,13 @@
-#define AUDIO_HEAP_SIZE 0x3800000 // The original audio heap size was too small. The heap would exceed capacity and corrupt everything in its path.
+#if defined(_WIN64) || defined(__x86_64__) || defined(__arm64__)
+#define _SOH64
+#define AUDIO_HEAP_SIZE 0x3800000
```
are these supposed to be the same?

louist103 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -589,64 +612,90 @@ extern "C" SequenceData ResourceMgr_LoadSeqByName(const char* path)
-        *strem++; // RIFF
-        *strem++; // size
-        *strem++; // WAVE
+    if (entry.ext == "wav") {
```
We should probably not rely on the file extension and use the actual header in the file

KiritoDv replied:
We talked about this Jack and me, tbh i prefer using extensions instead of reading the header

louist103 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -717,7 +766,7 @@ extern "C" SoundFont* ResourceMgr_LoadAudioSoundFont(const char* path) {
                 drum->envelope = (AdsrEnvelope*)malloc(sizeof(AdsrEnvelope) * soundFont->drums[i].env.size());
 
-                for (int k = 0; k < soundFont->drums[i].env.size(); k++) 
+                for (int k = 0; k < soundFont->drums[i].env.size(); k++)
```
Either make this ranged based or use `size_t`

louist103 commented on soh/src/code/audio_load.c:
```diff
@@ -1399,7 +1394,7 @@ void AudioLoad_Init(void* heap, u32 heapSize) {
     }
 
-    for (i = 0; i < (s32)gAudioContext.audioHeapSize / 8; i++) {
+    for (i = 0; i < gAudioContext.audioHeapSize / 8; i++) {
```
Why is `memset` used above and its manually zeroed out here?

louist103 commented on soh/src/code/audio_load.c:
```diff
@@ -1450,7 +1445,7 @@ void AudioLoad_Init(void* heap, u32 heapSize) {
         char** seqList = ResourceMgr_ListFiles("audio/sequences*", &seqListSize);
 
-        for (int i = 0; i < seqListSize; i++) 
+        for (int i = 0; i < seqListSize; i++)
```
Again with the signed iterator.

Kenix3 replied:
```suggestion
        for (size_t i = 0; i < seqListSize; i++)
```

louist103 commented on soh/src/overlays/misc/ovl_kaleido_scope/z_kaleido_scope_PAL.c:
```diff
@@ -3159,11 +3159,11 @@ void KaleidoScope_Update(GlobalContext* globalCtx)
-    size_t size1;
-    size_t size2;
-    u16 i;
+    size_t size = 0;
+    size_t size0 = 0;
+    size_t size1 = 0;
+    size_t size2 = 0;
+    u16 i = 0;
```
Is this needed?

NEstelami replied:
I'm guessing MSVC was complaining about uninitialized stack variables.

KiritoDv committed:
Added fixes based on the reviews

NEstelami commented on soh/src/code/audio_init_params.c:
```diff
@@ -10,8 +10,11 @@ const s16 D_8014A6C0[] = {
 
-//const AudioContextInitSizes D_8014A6C4 = { 0x37F00, 0xE0E0, 0xBCE0 };
-const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 }; // OTRTODO: This might be overkill...
+#ifdef _SOH64  // OTRTODO: This might be overkill...
+const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 };
+#else
+const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 };
+#endif
```
Lines 14 and 16 are exactly the same. Is the ifdef necessary?

KiritoDv replied:
Im not sure, you increased the size on 32 bits, thats why im not sure what to do

NEstelami replied:
I would just remove the if def and only have line 14. Those sizes are more than enough for both 32 and 64 bit.

NEstelami commented on libultraship/libultraship/DisplayList.cpp:
```diff
@@ -12,20 +12,51 @@ namespace Ship
+
+				if (opcode == G_ENDDL)
+					break;
+			}
```
What is the reasoning for having two different implementations here?

KiritoDv replied:
If im not bad, Emil increased the size of the dlists on 64 bits

NEstelami replied:
DLists in RAM were increased in size, yes. But the DLists in the OTR are always 64-bits for each `Gfx`. The in-memory DLists being 128-bits per `Gfx` just means there's more room for 64-bit pointers. I don't think we need to split the 64-bits into two and treat them as two `Gfx`?

Kenix3 replied:
@NEstelami Good with these display list changes after recent updates?

Kenix3 replied:
CC @Emill

KiritoDv committed:
Removed ifdefs for audio heap sizes

dcvz committed:
Extract StormLib building from soh to ZAPDTR (#505)

* Extract StormLib building from soh to ZAPDTR

ZAPDTR is currently where we have everything… though this might not make the most sense.

* Do not import storm as a relative

KiritoDv committed:
Fixed audio_heap_size

Kenix3 committed:
Update soh/src/code/audio_load.c

Kenix3 commented on soh/src/code/audio_init_params.c:
```diff
@@ -10,8 +10,11 @@ const s16 D_8014A6C0[] = {
+const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 };
+#else
+const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 };
+#endif
```
```suggestion
const AudioContextInitSizes D_8014A6C4 = { 0x37F000, 0xE0E00, 0xBCE00 };
```

Kenix3 committed:
Merge branch 'develop' into zelda64

dcvz committed:
Jenkins: Fix path to soh.exe
